### PR TITLE
polymorphic muxing (for social demo)

### PR DESCRIPTION
**Pull request by mmandlis**
support different templates be used for hosted slots.

moved templates caching to DomContext.
templates are now cached by:
`particleName::slotName::templateName // currently templateName is always 'default'`
or for hosted templates:
`particleName::slotName::hostedParticleName::hostedSlotName::templateName`

next step is to either only request template once per hosted particle class.

### CODE REVIEW

File: runtime/dom-set-context.js [hunk 1]
```diff
@@ -41,6 +41,22 @@ export class DomSetContext {
     return Object.keys(this._contextBySubId).length == Object.keys(context).length &&
            !Object.keys(this._contextBySubId).find(c => this._contextBySubId[c] != context[c]);
   }
+  setTemplate(templatePrefix, templateName, template) {
+    let isStringTemplateName = typeof template == 'string';
```
Comment (shaper) [inline runtime/dom-set-context.js]: I think this was intended to be `templateName` not `template` on the RHS?

File: runtime/dom-slot.js [hunk 2]
```diff
@@ -133,13 +113,31 @@ export class DomSlot extends Slot {
...
       request.push('template');
     }
     return request;
   }
+  formatHostedContent(hostedSlot, content) {
+    if (content.templateName) {
+      if (typeof content.templateName == 'string') {
+        content.templateName = `${hostedSlot.particleName}::${hostedSlot.slotName}::${content.templateName}`;
+      } else {
+        assert(false, 'TODO: Implement this!');
```
Comment (shaper) [inline runtime/dom-slot.js]: What is this case, is it for when we pass the model, or a template already generated from a string? Perhaps add comment to note.

File: runtime/test/dom-slot-tests.js [hunk 3]
```diff
@@ -10,27 +10,36 @@
...
     this.context = context;
   }
   createTemplateElement(template) {
     return template;
   }
+  setTemplate(templatePrefix, templateName, template) {
+    assert(template);
+    templates[`${templatePrefix}::${templateName}`] = template;
+  }
+  hasTemplate(templateName) { return Object.keys(templates).find(key => key.startsWith(templateName)); } //!!/*this.*/templates[templateName]; }
```
Comment (shaper) [inline runtime/test/dom-slot-tests.js]: Probably don't need the commented out bit any more?

File: runtime/transformation-dom-particle.js [hunk 4]
```diff
@@ -21,6 +21,9 @@ export class TransformationDomParticle extends DomParticle {
   getTemplate(slotName) {
     return this._state.template;
   }
+  getTemplateName(slotName) {
+    return this._state.templateName;
```
Comment (shaper) [inline runtime/transformation-dom-particle.js]: We never need to include the slot name for the base TransformationDomParticle? May be worth a comment to explain.

File: runtime/dom-slot.js (hunk 2 above)
Comment (mmandlis) [reply runtime/dom-slot.js]: added comment (this for set slots rendered by hosted particle)

**Comment by mmandlis** [comment]
Filed https://github.com/PolymerLabs/arcs/issues/1360 to write a design doc for this whole thing.
Walter, I'll be happy to work on this together (next week, or any time)

**Pull request merged**